algorand/go-algorand
# go-algorand 2.5.1-beta

![GitHub Logo](https://www.algorand.com/static/link-grid-bottom-252785b210330feb4966f54fc17cc7cc.jpg)

## Highlights

Support for TEAL 3
- Added support for the following new opcodes:
  - assert
  - min_balance
  - getbit
  - setbit
  - getbyte
  - setbyte
  - swap
  - select
  - dig
  - stxn
  - stxna
  - pushbytes
  - pushint
- Added support for the following txn fields:
  - ForeignAssets
  - NumForeignAssets
  - ForeignApps
  - NumForeignApps
  - GlobalStateInts
  - GlobalStateByteslices
  - LocalStateInts
  - LocalStateByteslices
- Added support for new global field:
  - CreatorAddress

Also includes Merkle root transaction hash and initial rewards calculation fix.


**This release will require a protocol upgrade.**


## Changelog

1. TEAL
    - New Feature - TEAL 3 support (see Highlights)
1. Agreement
    - New Feature - add a Merkle commitment to transactions in block header
    - Bug Fix - fix initial rewards rate calculation for new networks
1. Network
    - Enhancement
        - Network: refactor broadcastThread
        - Remove the network v1 fetcher via websocket connections
        - Deprecate support of reverse catchup protocol ( part 2 )
        - Deprecate support of reverse catchup protocol ( part 3 )
1. Logging
    - Bug Fixes
        - Handle #pragma errors more like all others
        - Prevent misleading agreement error message
1. Tools
    - Enhancement - add appeal to kindness to dispenser to help prevent abuse
1. Tests
    - Enhancements
        - Put some unicode in an AsssetName to help test Indexer
        - Improve the reliability of the TestApplicationsUpgradeOverGossip e2e test by using a specific historical protocols and avoiding dynamically modifying current and/or future protocols.
        - Remove GOPATH dependencies in integration tests
        - Remove data directory from start/stop network expect tests
1. Other
    - Enhancements
        - Clarify account status meaning in code-comment
        - Fix few minor typos
        - Integrate package signing into CI pipeline
        - Make github templates friendlier
        - Add comment details on format of Transaction for kmd

## Protocol Upgrade

This release contains a consensus protocol upgrade, which implements the following spec:
https://github.com/algorandfoundation/specs/tree/ac2255d586c4474d4ebcf3809acccb59b7ef34ff

## Additional Resources
* [Algorand Forum](https://forum.algorand.org)
* [Developer Documentation](https://developer.algorand.org)

## Test Plan

Standard release testing.
Consensus upgrade testing on BetaNet.
